File: dashboard/lib/monitor/providers/digitalocean.ts
import { httpGet, prorated } from "../util";
import type { Row } from "../types";

const API = "https://api.digitalocean.com/v2";
const VOL_GB_MONTH = 0.1;
const LB_MONTHLY = 12;

async function paged<T>(url: string, h: Record<string, string>): Promise<{ kind: string; item: T }[]> {
  const out: { kind: string; item: T }[] = [];
  let next: string | null = url;
  while (next) {
    const data: {
      droplets?: T[]; volumes?: T[]; load_balancers?: T[];
      links?: { pages?: { next?: string } };
    } = await httpGet(next, h);
    for (const k of ["droplets", "volumes", "load_balancers"] as const) {
      for (const item of data[k] || []) out.push({ kind: k, item });
    }
    next = data.links?.pages?.next || null;
  }
  return out;
}

export async function collect(account: string, token: string): Promise<Row[]> {
  const h = { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };
  const rows: Row[] = [];
  for (const { kind, item } of await paged<Record<string, never>>(`${API}/droplets?per_page=200`, h)) {
    const o = item as unknown as {
      id: number; name: string; status: string; created_at: string; tags?: string[];
      region?: { slug: string }; size?: { slug: string; price_monthly: number };
    };
    if (kind === "droplets") {
      rows.push({
        provider: "digitalocean", account, service: "droplet", resource_id: String(o.id),
        name: o.name, region: o.region?.slug || "", status: o.status,
        cost_mtd: prorated(o.created_at, o.size?.price_monthly || 0),
        tags: { tags: o.tags || [] }, raw: { size: o.size?.slug },
      });
    } else if (kind === "volumes") {
      const v = o as unknown as { id: string; name: string; size_gigabytes: number; region?: { slug: string } };
      rows.push({
        provider: "digitalocean", account, service: "volume", resource_id: v.id,
        name: v.name, region: v.region?.slug || "", status: "",
        cost_mtd: Math.round(v.size_gigabytes * VOL_GB_MONTH * 10000) / 10000,
        tags: {}, raw: { size_gb: v.size_gigabytes },
      });
    } else if (kind === "load_balancers") {
      const lb = o as unknown as { id: string; name: string; status: string; created_at: string; region?: { slug: string } };
      rows.push({
        provider: "digitalocean", account, service: "load_balancer", resource_id: lb.id,
        name: lb.name, region: lb.region?.slug || "", status: lb.status,
        cost_mtd: prorated(lb.created_at, LB_MONTHLY), tags: {}, raw: {},
      });
    }
  }
  return rows;
}

File: monitor/digitalocean.py
"""DigitalOcean (api.digitalocean.com/v2). Bearer token. Docs: docs.digitalocean.com/reference/api."""
from __future__ import annotations

from .base import http_get, mtd_factor, parse_time

API = "https://api.digitalocean.com/v2"
VOL_GB_MONTH = 0.10
LB_MONTHLY = 12.0


def _headers(token):
    return {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}


def _paged(url, headers):
    out, next_url = [], url
    while next_url:
        data = http_get(next_url, headers)
        for key in ("droplets", "volumes", "load_balancers", "domains", "snapshots"):
            out.extend([(key, o) for o in data.get(key, [])])
        links = (data.get("links") or {}).get("pages") or {}
        next_url = (links.get("next") or "") or None
    return out


def collect(account, token):
    h = _headers(token)
    rows = []
    for kind, o in _paged(f"{API}/droplets?per_page=200", h):
        if kind != "droplets":
            continue
        monthly = float(((o.get("size") or {}).get("price_monthly") or 0))
        rows.append({
            "provider": "digitalocean", "account": account, "service": "droplet",
            "resource_id": str(o["id"]), "name": o.get("name", ""),
            "region": (o.get("region") or {}).get("slug", ""),
            "status": o.get("status", ""),
            "cost_mtd": round(monthly * mtd_factor(parse_time(o.get("created_at"))), 4),
            "tags": {"tags": o.get("tags", [])}, "raw": {"size": (o.get("size") or {}).get("slug")},
        })
    for kind, o in _paged(f"{API}/volumes?per_page=200", h):
        if kind != "volumes":
            continue
        monthly = float(o.get("size_gigabytes", 0)) * VOL_GB_MONTH
        rows.append({
            "provider": "digitalocean", "account": account, "service": "volume",
            "resource_id": o.get("id", ""), "name": o.get("name", ""),
            "region": (o.get("region") or {}).get("slug", ""),
            "status": "", "cost_mtd": round(monthly, 4),
            "tags": {}, "raw": {"size_gb": o.get("size_gigabytes")},
        })
    for kind, o in _paged(f"{API}/load_balancers?per_page=200", h):
        if kind != "load_balancers":
            continue
        rows.append({
            "provider": "digitalocean", "account": account, "service": "load_balancer",
            "resource_id": o.get("id", ""), "name": o.get("name", ""),
            "region": (o.get("region") or {}).get("slug", ""),
            "status": o.get("status", ""),
            "cost_mtd": round(LB_MONTHLY * mtd_factor(parse_time(o.get("created_at"))), 4),
            "tags": {}, "raw": {},
        })
    return rows
